core/trino-main/src/main/java/io/trino/execution/scheduler/AllAtOnceExecutionSchedule.java

```diff
@@ -99,10 +96,9 @@ public boolean isFinished()
         Set<PlanFragment> rootFragments = fragments.stream()
                 .filter(fragment -> !remoteSources.contains(fragment.getId()))
                 .collect(toImmutableSet());
-        checkArgument(rootFragments.size() == 1, "Expected one root fragment, but found: %s", rootFragments);
 
         Visitor visitor = new Visitor(fragments);
-        visitor.processFragment(getOnlyElement(rootFragments).getId());
+        rootFragments.forEach(fragment -> visitor.processFragment(fragment.getId()));
```

Comment:
Can you explain why this changed? What does it mean for there to be multiple root fragments?

Reply:
Currently coordinator stages are scheduled separately as those are never restarted. Coordinator stage may have multiple source stages, that now are also scheduled separately. Thus the `ExecutionSchedule` must be able to schedule a graph of stages with multiple roots.